File: app/schemas/transaction.py

# app/schemas/transaction.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.transaction import StatusSolicitacao, TipoMovimentacao
from app.schemas.asset import AssetResponse
from app.schemas.user import UserResponse


# ----- Movimentacao -----
class MovimentacaoBase(BaseModel):
    asset_id: int
    tipo: TipoMovimentacao
    de_user_id: int | None = None
    para_user_id: int | None = None
    de_departamento_id: int | None = None
    para_departamento_id: int | None = None
    observacao: str | None = None

class MovimentacaoCreate(MovimentacaoBase):
    pass

class MovimentacaoResponse(MovimentacaoBase):
    id: int
    data: datetime
    de_user: UserResponse | None = None
    para_user: UserResponse | None = None
    asset: AssetResponse | None = None

    model_config = ConfigDict(from_attributes=True)

# ----- Solicitacao -----
class SolicitacaoBase(BaseModel):
    asset_id: int | None = None
    motivo: str
    data_prevista_devolucao: datetime | None = None

class SolicitacaoCreate(SolicitacaoBase):
    pass

class SolicitacaoUpdate(BaseModel):
    status: StatusSolicitacao | None = None
    aprovador_id: int | None = None
    data_aprovacao: datetime | None = None

class SolicitacaoResponse(SolicitacaoBase):
    id: int
    solicitante_id: int
    status: StatusSolicitacao
    data_solicitacao: datetime
    
    aprovador_id: int | None = None
    data_aprovacao: datetime | None = None
    
    solicitante: UserResponse | None = None
    aprovador: UserResponse | None = None
    asset: AssetResponse | None = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/transaction.ts
import type { Asset } from './asset';

export interface Solicitacao {
  id: number;
  solicitante_id?: number;
  solicitante?: {
    id: number;
    nome: string;
    email: string;
  };
  asset_id?: number;
  asset?: Asset;
  motivo: string;
  status: 'pendente' | 'aprovada' | 'rejeitada' | 'entregue' | 'devolvida';
  data_solicitacao: string;
  data_aprovacao?: string;
  aprovador_id?: number;
  aprovador?: {
    id: number;
    nome: string;
    email: string;
  };
  data_entrega?: string;
  confirmado_por_id?: number;
  confirmador?: {
    id: number;
    nome: string;
    email: string;
  };
  confirmado_via_qr?: boolean;
  observacao_entrega?: string;
  data_devolucao?: string;
  recebido_por_id?: number;
  recebedor?: {
    id: number;
    nome: string;
    email: string;
  };
  condicao_devolucao?: string;
  acessorios_devolvidos?: string;
  observacoes_devolucao?: string;
  observacao_devolucao?: string;
  data_prevista_devolucao?: string;
}

export interface Movimentacao {
  id: number;
  asset_id: number;
  asset?: Asset;
  tipo: 'entrada' | 'saida' | 'transferencia' | 'emprestimo' | 'devolucao' | 'manutencao' | 'baixa';
  data_movimentacao: string;
  de_user_id?: number;
  de_user?: {
    id: number;
    nome: string;
    email: string;
  };
  para_user_id?: number;
  para_user?: {
    id: number;
    nome: string;
    email: string;
  };
  de_local_id?: number;
  para_local_id?: number;
  observacao?: string;
}
